src/core/ext/transport/chttp2/transport/hpack_parser.cc

```diff
@@ -1494,7 +1495,7 @@ static grpc_error* parse_key_string(grpc_chttp2_hpack_parser* p,
 /* check if a key represents a binary header or not */
 
 static bool is_binary_literal_header(grpc_chttp2_hpack_parser* p) {
-  return grpc_is_binary_header(
+  return grpc_is_binary_header_refcounted(
```

Comment:
at the places where these are used, how do we know that the slices are refcounted for sure?

Reply:
Two cases: is_binary_literal_header, and is_binary_indexed_header.

In the indexed case: elem is a result of grpc_chttp2_hptbl_lookup. That means either elem was in the static entries (which have a refcount) or they were added via grpc_chttp2_hptbl_add, which is only performed if the elem was static or intern specified storage - in both cases, the key has a refcount.

In the literal case: either we look at the output of slice_from_static_buffer (has a refcount) or p->key.data.referenced. p->key.data.referenced is either the output of grpc_empty_slice() (NULL refcount!) or it has a refcount set in begin_parse_string(). The act of setting that refcount determines which of the two actions we take; if we set the refcount, we use p->key.data.referenced, which is guaranteed to have a non-null refcount. If we do not set the refcount (in which case, sure, p->key.data.referenced could be grpc_empty_slice()) then we do not read it anyways. So in either case the refcount is set.

So in both places it is refcounted.

Reply:
A few things:

1) Are we sure that it's refcounted because of the way the code happens to be implemented right now or because it wouldn't logically make sense otherwise? I am concerned about making a change that is fragile to future implementation changes.

2) Can you add your reasoning in your comment above as a comment somewhere? Maybe above the definition of grpc_is_binary_header_refcounted?

3) From the name "grpc_is_binary_header_refcounted", one could get confused and think that it's checking whether the binary header is refcounted. Perhaps rename to "grpc_is_refcounted_slice_binary_header"

Reply:
Ack on adding the comment and renaming the method -  will do both shortly.

Re: the implementation: for the indexed case I believe that it will be always be refcounted - assuming we're indexing in the first place for cheaper slice operations I think it will be unlikely that externally allocated or inlined slices would make it into an index table.

For the literal case: right now the parser implementation dictates that it's always a refcounted slice. I think that in this case, maybe it could change in the future; that said, if it ever did, I think having a debug assert in our fastpath method would make it very obvious that our prior assumptions no longer held.

At that point, it would be easy enough to change it back to the more general method that can take either kind. I don't think it's fragile in the sense that if we ever changed it to break, we'd immediately catch the break - so it's not like we'd be introducing a heisenbug.

Reply:
Added comments describing why slices are refcounted, per discussion with @yashykt and @hcaseyal . Changed method name per @hcaseyal .